javascript/ql/src/semmle/javascript/Regexp.qll

```diff
@@ -404,3 +407,87 @@ module RegExpPatterns {
     result = "(?:com|org|edu|gov|uk|net|io)(?![a-z0-9])"
   }
 }
+
+/**
+ * Gets a node whose value may flow (inter-procedurally) to a position where it is interpreted
+ * as a part of a regular expression.
+ */
+private DataFlow::Node regExpSource(DataFlow::Node re, DataFlow::TypeBackTracker t) {
+  t.start() and
+  re = result and
+  isInterpretedAsRegExp(result)
+  or
+  exists(DataFlow::TypeBackTracker t2, DataFlow::Node succ | succ = regExpSource(re, t2) |
+    t2 = t.smallstep(result, succ)
+    or
+    any(TaintTracking::AdditionalTaintStep dts).step(result, succ) and
+    t = t2
+  )
+}
+
+/**
+ * Gets a node whose value may flow (inter-procedurally) to a position where it is interpreted
+ * as a part of a regular expression.
+ */
+private DataFlow::Node regExpSource(DataFlow::Node re) {
+  result = regExpSource(re, DataFlow::TypeBackTracker::end())
+}
+
+/**
+ * A node whose value may flow to a position where it is interpreted
+ * as a part of a regular expression.
+ */
+abstract class RegExpPatternSource extends DataFlow::Node {
+  /**
+   * Gets the pattern of this node.
+   */
+  abstract string getPattern();
+
+  /**
+   * Gets a regular expression object that is built from the value of this node.
+   */
+  abstract DataFlow::SourceNode getARegExpObject();
+}
+
+/**
+ * A regular expression literal, viewed as the pattern source for itself.
+ */
+class RegExpLiteralPatternSource extends RegExpPatternSource {
```

Comment:
Could/should this class be `private`?

Reply:
For consistency witth the public `StringRegExpPatternSource`, I think this should remain public

Reply:
Hm, fine. I don't like having all these awkwardly named classes in the global namespace.

Reply:
I can hide them in a `module BuiltinRegExpPatternSources { class RegExpLiteralPatternSource ... }`, that is half the awkwardness in the global namespace.

Reply:
No, I don't think that would be better. Could the above two uses of `StringRegExpPatternSource` be encapsulated as member predicates of `RegExpPatternSource`?

Reply:
I have introduced `getAParse` instead of `getAUse`, then the predicate name makes sense for the regular expression literals which otherwise should be flagged at all their usages. As an added bonus, the message in `js/incomplete-hostname-regexp` will no longer include a trivial link for cases where the string literal already is at a location where it is used as a regular expression:

That is, the message for:
```
new RegExp(`test.example.com$`); // NOT OK
```
is now:
```
This regular expression has an unescaped '.' before 'example.com', so it might match more hosts than expected. 
```